### PR TITLE
feat: add initGPUInfoCM func

### PR DESCRIPTION
**Pull request by nowinkeyy**
1. add node-gpu-info configmap & role & rolebinding to deploy.yaml
2. use clientSet to get resources from the API Server, not from Informer's cache

### CODE REVIEW

File: controllers/node/deploy/Kubefile [hunk 1]
```diff
@@ -7,4 +7,4 @@ COPY manifests manifests
 
 ENV DEFAULT_NAMESPACE node-system
 
-CMD ["kubectl apply -f manifests/deploy.yaml -n $DEFAULT_NAMESPACE"]
+CMD ["kubectl apply -f manifests/deploy.yaml -n $DEFAULT_NAMESPACE --env=DEFAULT_NAMESPACE=node-system"]
```
Comment (bxy4543) [inline controllers/node/deploy/Kubefile:10]: `--env=DEFAULT_NAMESPACE=node-system` ?

File: controllers/node/controllers/gpu_controller.go [hunk 2]
```diff
@@ -137,7 +145,13 @@ func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Res
...
 	// create or update gpu-info configmap
 	configmap := &corev1.ConfigMap{}
-	err = r.Get(ctx, types.NamespacedName{Name: GPUInfo, Namespace: GPUInfoNameSpace}, configmap)
+
+	if clientSet != nil {
+		configmap, err = clientSet.CoreV1().ConfigMaps(GPUInfoNameSpace).Get(ctx, GPUInfo, metaV1.GetOptions{})
+	} else {
+		err = r.Get(ctx, types.NamespacedName{Name: GPUInfo, Namespace: GPUInfoNameSpace}, configmap)
+	}
+
```
Comment (lingdie) [inline controllers/node/controllers/gpu_controller.go:154]: do not handle error like this.

File: controllers/node/controllers/gpu_controller.go [hunk 3]
```diff
@@ -153,6 +167,9 @@ func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Res
 			return ctrl.Result{}, err
 		}
 	} else if err == nil {
+		if configmap.Data == nil {
```
Comment (lingdie) [inline controllers/node/controllers/gpu_controller.go:170]: use err != nil or just delete `else if`

File: controllers/node/controllers/gpu_controller.go [hunk 4]
```diff
@@ -164,15 +181,58 @@ func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Res
...
+func (r *GpuReconciler) initGPUInfoCM(clientSet *kubernetes.Clientset) error {
+	// filter for nodes that have GPU
+	req1, _ := labels.NewRequirement(NvidiaGPUProduct, selection.Exists, []string{})
+	req2, _ := labels.NewRequirement(NvidiaGPUMemory, selection.Exists, []string{})
+	selector := labels.NewSelector().Add(*req1, *req2)
+	listOpts := metaV1.ListOptions{
+		LabelSelector: selector.String(),
+	}
+
+	nodeList, err := clientSet.CoreV1().Nodes().List(context.TODO(), listOpts)
```
Comment (lingdie) [inline controllers/node/controllers/gpu_controller.go:198]: use reconciler.context.

File: controllers/node/controllers/gpu_controller.go [hunk 5]
```diff
@@ -164,15 +181,58 @@ func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Res
...
+		list, err := clientSet.CoreV1().Pods("").List(context.TODO(), metaV1.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector("spec.nodeName", item.Name).String(),
+		})
+		if err != nil {
+			return err
+		}
+		podList.Items = append(podList.Items, list.Items...)
+	}
+
+	_, err = r.applyGPUInfoCM(context.TODO(), nodeList, podList, clientSet)
```
Comment (lingdie) [inline controllers/node/controllers/gpu_controller.go:214]: no need to new a context.

File: controllers/node/controllers/gpu_controller.go (hunk 3 above)
Comment (nowinkeyy) [reply controllers/node/controllers/gpu_controller.go:170]: is it ok?
image

**Review by lingdie — APPROVED** [review]

**Review by bxy4543 — APPROVED** [review]

**Pull request merged**